src/nupic/algorithms/TemporalMemory.cpp

```diff
@@ -0,0 +1,1093 @@
+
+    tie(bestCell, bestSegment) =
+      bestMatchingCell(cells, prevActiveCells, _connections);
+
+    if (bestCell) _winnerCells.insert(*bestCell);
+
+    if (bestSegment == NULL && prevWinnerCells.size())
+    {
+      bestSegment = new Segment(_connections.createSegment(*bestCell));
+    }
+
+    if (bestSegment != NULL)
+    {
+      _learningSegments.push_back(*bestSegment);
+    }
+  }
+
+  return make_tuple(_activeCells, _winnerCells, _learningSegments);
+}
+
+void TemporalMemory::learnOnSegments(
+  vector<Segment>& prevActiveSegments,
+  vector<Segment>& learningSegments,
+  set<Cell>& prevActiveCells,
+  set<Cell>& winnerCells,
+  set<Cell>& prevWinnerCells,
+  Connections& _connections)
+{
+  vector<Segment> _allSegments;
+
+  for (auto segment : prevActiveSegments)
+    _allSegments.push_back(segment);
+  for (auto segment : learningSegments)
+    _allSegments.push_back(segment);
+
+  for (Segment segment : _allSegments)
+  {
+    bool isLearningSegment = (find(
+      learningSegments.begin(), learningSegments.end(),
+      segment) != learningSegments.end());
+
+    bool isFromWinnerCell = (find(
+      winnerCells.begin(), winnerCells.end(),
+      segment.cell) != winnerCells.end());
+
+    vector<Synapse> activeSynapses(activeSynapsesForSegment(
+      segment, prevActiveCells, _connections));
+
+    if (isLearningSegment || isFromWinnerCell)
+    {
+      adaptSegment(segment, activeSynapses, _connections);
+    }
+
+    if (isLearningSegment)
+    {
+      Int n = maxNewSynapseCount_ - Int(activeSynapses.size());
+
+      set<Cell> cells = pickCellsToLearnOn(
+                          n, segment, prevWinnerCells, _connections);
+      for (Cell presynapticCell : cells)
+      {
+        _connections.createSynapse(segment, presynapticCell, initialPermanence_);
+      }
+    }
+  }
+}
+
+tuple<vector<Segment>, set<Cell>> TemporalMemory::computePredictiveCells(
+  set<Cell>& _activeCells,
+  Connections& _connections)
+{
+  map<Segment, Int> _numActiveConnectedSynapsesForSegment;
+
+  vector<Segment> _activeSegments;
+  set<Cell> _predictiveCells;
+
+  for (Cell cell : _activeCells)
+  {
+    for (Synapse synapse : _connections.synapsesForPresynapticCell(cell))
+    {
+      SynapseData synapseData = _connections.dataForSynapse(synapse);
+      Segment segment = synapse.segment;
+      Real permanence = synapseData.permanence;
+
+      if (permanence >= connectedPermanence_)
+      {
+        _numActiveConnectedSynapsesForSegment[segment] += 1;
+
+        if (_numActiveConnectedSynapsesForSegment[segment] >=
+          activationThreshold_)
+        {
+          _activeSegments.push_back(segment);
+          _predictiveCells.insert(Cell(segment.cell.idx));
+        }
+      }
+    }
+  }
+
+  return make_tuple(_activeSegments, _predictiveCells);
+}
+
+// ==============================
+//  Helper functions
+// ==============================
+
+tuple<Cell*, Segment*>
+TemporalMemory::bestMatchingCell(
```

Comment:
This could use `Connections.mostActiveSegmentForCells` instead of reimplementing it.